### PR TITLE
[netcore] Handle NATIVE_DLL_SEARCH_DIRECTORIES for p/invoke

### PR DESCRIPTION
**Pull request by filipnavara**
Extracted the good parts out of https://github.com/mono/mono/pull/12613 and fixed up enough so it does not crash.

### CODE REVIEW

**Comment by filipnavara** [comment]
I'm not entirely convinced this is the correct approach. I also get a lot of load attempts like this:
```
Mono: netcore DllImport fallback handler: wanted '/Users/filipnavara/Documents/mono/netcore/shared/Microsoft.NETCore.App/3.0.0-preview4-27514-06/System.Private.CoreLib.dll.dylib'
Mono: netcore DllImport fallback handler: wanted '/usr/local/lib/mono/aot-cache/amd64/System.Private.CoreLib.dll.dylib'
```
...which suggests some funky fallback logic and also that the code paths get hit for managed assemblies.

**Comment by lambdageek** [comment]
> which suggests that the code paths get hit for managed assemblies when looking up the AOT versions
Yea this is a little bit of an abuse of the DL fallback handler. It's used for all native libs, including AOT images.  I think it's more for platform support than for Mono itself.

It might be worth seeing if we can just have a netcore version of `pinvoke_probe_for_module` that uses the native paths from the host, instead of using the DL fallback.

**Comment by filipnavara** [comment]
> It might be worth seeing if we can just have a netcore version of pinvoke_probe_for_module that uses the native paths from the host, instead of using the DL fallback.

I'll check how much effort would that be. Probably not too much.

**Comment by filipnavara** [comment]
For reference, here's the CoreCLR algorithm: https://github.com/dotnet/coreclr/blob/2f4a7564beb6f8c1f4929504ee2221953d25fc39/src/vm/dllimport.cpp#L6570-L6777

The basic structure is similar to what Mono does, but details differ. They also loop over prefix/suffix variations and each iteration they run inner loop over possible paths. I am not sure if Mono handles "libNAME.so.VERSION" the same way they do. Some further aspects can be configured using the `DefaultDllImportSearchPaths` attribute, which is not implemented in Mono for sure.

I will try to move the NATIVE_DLL_SEARCH_DIRECTORIES logic to `pinvoke_probe_for_module_relative_directories`. Rest can be implemented later.

File: mono/metadata/loader.c [hunk 1]
```diff
@@ -1595,10 +1609,41 @@ pinvoke_probe_for_module_relative_directories (MonoImage *image, const char *fil
...
+			}
+			g_free (full_name);
+		}
 
+		if (j >= pinvoke_search_directories_count) {
+			g_free (mdirname);
+		}
+	}
+#else
 			for (j = 0; j < 3; ++j) {
```
Comment (filipnavara) [inline mono/metadata/loader.c]: The indentation is off here. I intentionally left it that way to keep the diff tidy, but I can fix it on request.

**Comment by filipnavara** [comment]
> Might be worth refactoring the code so that the inner while loop is a separate helper function. 

Agreed, will do.

**Comment by filipnavara** [comment]
Rebased on top of master to fix conflicts.

**Comment by filipnavara** [comment]
@marek-safar Friendly ping on this one. I have to rebase it almost every time `netcore/Makefile[.am]` gets changed. It removes one of the hacks necessary to run the `netcore` samples / unit tests.

**Pull request merged**